java/tracing/src/main/scala/org/typelevel/otel4s/java/trace/SpanBuilderImpl.scala

```diff
@@ -0,0 +1,279 @@
+
+          result <- Resource.make(
+            child("acquire", rootCtx).surround(resource.allocated)
+          ) { case (_, release) =>
+            child("release", rootCtx).surround(release)
+          }
+
+          useSpanBackend <- child("use", rootCtx)
+        } yield Span.Res.fromBackend(result._1, useSpanBackend)
+
+      case None =>
+        resource.map(a => Span.Res.fromBackend(a, Span.Backend.noop))
+    }
+  }
+
+  private def startManaged(
+      builder: JSpanBuilder,
+      timestampSelect: TimestampSelect
+  ): Resource[F, SpanBackendImpl[F]] = {
+
+    def acquire: F[SpanBackendImpl[F]] =
+      startSpan(builder, timestampSelect)
+
+    def release(backend: Span.Backend[F], ec: Resource.ExitCase): F[Unit] = {
+      def end: F[Unit] =
+        timestampSelect match {
+          case TimestampSelect.Explicit =>
+            for {
+              now <- Sync[F].realTime
+              _ <- backend.end(now)
+            } yield ()
+
+          case TimestampSelect.Delegate =>
+            backend.end
+        }
+
+      for {
+        _ <- finalizationStrategy
+          .lift(ec)
+          .foldMapM(SpanFinalizer.run(backend, _))
+        _ <- end
+      } yield ()
+    }
+
+    for {
+      backend <- Resource.makeCase(acquire) { case (b, ec) => release(b, ec) }
+      _ <- scope.makeScope(backend.jSpan)
+    } yield backend
+  }
+
+  private def startSpan(
+      builder: JSpanBuilder,
+      timestampSelect: TimestampSelect
+  ): F[SpanBackendImpl[F]] =
+    for {
+      builder <- timestampSelect match {
+        case TimestampSelect.Explicit if startTimestamp.isEmpty =>
+          for {
+            now <- Sync[F].realTime
+          } yield builder.setStartTimestamp(now.length, now.unit)
+
+        case _ =>
+          Sync[F].pure(builder)
+      }
+      jSpan <- Sync[F].delay(builder.startSpan())
+    } yield new SpanBackendImpl(
+      jSpan,
+      WrappedSpanContext(jSpan.getSpanContext)
+    )
+
+  private def makeJBuilder(parent: JContext): JSpanBuilder = {
+    val b = jTracer
+      .spanBuilder(name)
+      .setAllAttributes(Conversions.toJAttributes(attributes))
+      .setParent(parent)
+
+    kind.foreach(k => b.setSpanKind(toJSpanKind(k)))
+    b.setAllAttributes(Conversions.toJAttributes(attributes))
+    startTimestamp.foreach(d => b.setStartTimestamp(d.length, d.unit))
+    links.foreach { case (ctx, attributes) =>
+      b.addLink(
+        WrappedSpanContext.unwrap(ctx),
+        Conversions.toJAttributes(attributes)
+      )
+    }
+
+    b
+  }
+
+  private def parentContext: F[Option[JContext]] =
+    parent match {
+      case Parent.Root =>
+        scope.current.flatMap {
+          case TraceScope.Scope.Root(ctx) =>
+            Sync[F].pure(Some(ctx))
+
+          case TraceScope.Scope.Span(_, _, _) =>
+            scope.root.map(s => Some(s.ctx))
+
+          case TraceScope.Scope.Noop =>
+            Sync[F].pure(None)
```

Comment:
Are we going to create enough of these to cache this value in the builder?

Reply:
`parentContext` should be executed only once upon starting a span via one of the methods: `createManual`, `createAuto`, `createRes`.